### PR TITLE
Fix GetPathRoot issues

### LINKED ISSUE #27552: Several FileSystem and Path tests failing with latest coreclr
```
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"C:\\\\\\\\\", expected: \"C:\\\\\") [FAIL]
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:                    ↓ (pos 3)
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:       Expected: C:\
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:       Actual:   C:\\
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:                    ↑ (pos 3)
2018-02-28 13:24:45,135: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"C:\\\\\\\\foo2\", expected: \"C:\\\\\") [FAIL]
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:                    ↓ (pos 3)
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:       Expected: C:\
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:       Actual:   C:\\
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:                    ↑ (pos 3)
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,244: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"\\\\\\\\test\\\\unc\\\\path\\\\to\\\\something\", expected: \"\\\\\\\\test\\\\unc\") [FAIL]
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                           ↓ (pos 10)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Expected: \\test\unc
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Actual:   \\test\unc\
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                           ↑ (pos 10)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"\\\\\\\\a\\\\b\\\\c\\\\d\\\\e\", expected: \"\\\\\\\\a\\\\b\") [FAIL]
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                      ↓ (pos 5)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Expected: \\a\b
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Actual:   \\a\b\
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                      ↑ (pos 5)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"\\\\\\\\a\\\\b\\\\\", expected: \"\\\\\\\\a\\\\b\") [FAIL]
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                      ↓ (pos 5)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Expected: \\a\b
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Actual:   \\a\b\
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                      ↑ (pos 5)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"\\\\\\\\?\\\\UNC\\\\test\\\\unc\\\\path\\\\to\\\\something\", expected: \"\\\\\\\\?\\\\UNC\\\\test\\\\unc\") [FAIL]
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                                 ↓ (pos 16)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Expected: \\?\UNC\test\unc
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Actual:   \\?\UNC\test\unc\
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:                                 ↑ (pos 16)
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,260: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Span(value: \"\\\\\\\\?\\\\UNC\\\\a\\\\b2\\\\\", expected: \"\\\\\\\\?\\\\UNC\\\\a\\\\b2\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                             ↓ (pos 12)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\?\UNC\a\b2
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\?\UNC\a\b2\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                             ↑ (pos 12)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.netcoreapp.cs(17,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Span(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(value: \"\\\\\\\\test\\\\unc\\\\path\\\\to\\\\something\", expected: \"\\\\\\\\test\\\\unc\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                           ↓ (pos 10)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\test\unc
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\test\unc\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                           ↑ (pos 10)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.cs(48,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(value: \"\\\\\\\\a\\\\b\\\\c\\\\d\\\\e\", expected: \"\\\\\\\\a\\\\b\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                      ↓ (pos 5)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\a\b
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\a\b\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                      ↑ (pos 5)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.cs(48,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(value: \"\\\\\\\\a\\\\b\\\\\", expected: \"\\\\\\\\a\\\\b\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                      ↓ (pos 5)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\a\b
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\a\b\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                      ↑ (pos 5)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.cs(48,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(value: \"\\\\\\\\?\\\\UNC\\\\test\\\\unc\\\\path\\\\to\\\\something\", expected: \"\\\\\\\\?\\\\UNC\\\\test\\\\unc\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                                 ↓ (pos 16)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\?\UNC\test\unc
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\?\UNC\test\unc\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                                 ↑ (pos 16)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.cs(48,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(String value, String expected)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(value: \"\\\\\\\\?\\\\UNC\\\\a\\\\b2\\\\\", expected: \"\\\\\\\\?\\\\UNC\\\\a\\\\b2\") [FAIL]
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                             ↓ (pos 12)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\?\UNC\a\b2
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\?\UNC\a\b2\
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:                             ↑ (pos 12)
2018-02-28 13:24:45,276: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:45,292: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.Runtime.Extensions\tests\System\IO\PathTests_Windows.cs(48,0): at System.IO.Tests.PathTests_Windows.GetPathRoot_Windows(String value, String expected)
```
and
```
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.FileInfo_Directory.UNCShares [FAIL]
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:                                    ↓ (pos 19)
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:       Expected: \\Machine\Directory
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:       Actual:   \\Machine\Directory\
2018-02-28 13:24:34,276: INFO: proc(54): run_and_log_output: Output:                                    ↑ (pos 19)
2018-02-28 13:24:34,292: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:34,401: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.IO.FileSystem\tests\FileInfo\Directory.cs(55,0): at System.IO.Tests.FileInfo_Directory.UNCShares()
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:    System.IO.Tests.FileInfo_DirectoryName.UNCShares [FAIL]
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:       Assert.Equal() Failure
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:                                    ↓ (pos 19)
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:       Expected: \\Machine\Directory
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:       Actual:   \\Machine\Directory\
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:                                    ↑ (pos 19)
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:       Stack Trace:
2018-02-28 13:24:38,369: INFO: proc(54): run_and_log_output: Output:          D:\j\workspace\windows-TGrou---74aa877a\src\System.IO.FileSystem\tests\FileInfo\Directory.cs(55,0): at System.IO.Tests.FileInfo_Directory.UNCShares()
```

### PR DESCRIPTION
**Pull request by JeremyKuhne**
cc: @stephentoub, @Anipik, @danmosemsft, @pjanotti 

fixes #27552

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
@dotnet-bot test Packaging All Configurations x64 Debug Build

```
12:51:02   Restoring packages for D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\ilasm\ilasm.depproj...
12:51:02 D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\dotnetcli\sdk\2.1.2\NuGet.targets(103,5): error : Unable to load the service index for source https://dotnet.myget.org/F/dotnet-core/api/v3/index.json. [D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\ilasm\ilasm.depproj]
12:51:02 D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\dotnetcli\sdk\2.1.2\NuGet.targets(103,5): error :   The HTTP request to 'GET https://dotnet.myget.org/F/dotnet-core/api/v3/index.json' has timed out after 100000ms. [D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\ilasm\ilasm.depproj]
12:51:02 
12:51:02 Build FAILED.
```

File: src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs [hunk 1]
```diff
@@ -188,64 +188,57 @@ internal static int GetRootLength(ReadOnlySpan<char> path)
...
-                    i = uncRootLength;
-                    int n = 2; // Maximum separators to skip
+                    // UNC (\\?\UNC\ or \\), scan past server\share
+
+                    // Start past the prefix ("\\" or "\\?\UNC\")
+                    i = deviceUnc ? UncExtendedPrefixLength : UncPrefixLength;
+
+                    // Skip two separators at most
+                    int n = 2;
                     while (i < pathLength && (!IsDirectorySeparator(path[i]) || --n > 0))
```
Comment (danmosemsft) [inline src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs]: I find this hard to read as normally you would expect the termination check to be &&. But I see what you are doing is if the first clause is false, only then does it decrement n.

```c#
                     while (i < pathLength && (!IsDirectorySeparator(path[i]) || --n > 0))
                         i++;	                        
```
is this as good?
```c#
                     while (i < pathLength && n > 0))
                        if (IsDirectorySeparator(path[i]))
                                   n--;
                         i++;	                        
```

**Comment by danmosemsft** [comment]
Consider an outerloop randomizer test for some of these low level API's that are heavy on indexing through strings? Just feeds garbalized combinations of \ / ? a-z etc to check they don't throw NRE.

File: src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs]: Yeah, this one I've never been happy with. Some legacy logic I've brought along. I'll give it some thought.

**Comment by JeremyKuhne** [comment]
One OSX failure. Nothing to do with this PR. I'm looking into what might have caused this failure.

```
System.IO.Tests.Enumeration.SkipAttributeTests
--
SkippingHiddenFiles

System.ArgumentException : Destination is too short.

   at System.ReadOnlySpan`1.CopyTo(Span`1 destination) in /Users/buildagent/agent/_work/482/s/src/mscorlib/shared/System/ReadOnlySpan.Fast.cs:line 174
   at System.IO.Enumeration.FileSystemEntry.get_FullPath()
   at System.IO.Enumeration.FileSystemEntry.ToFullPath()
   at System.IO.Tests.Enumeration.SkipAttributeTests.<>c.<GetPaths>b__0_0(FileSystemEntry& entry) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.IO.FileSystem/tests/Enumeration/SkipAttributeTests.netcoreapp.cs:line 17
   at System.IO.Enumeration.FileSystemEnumerable`1.DelegateEnumerator.TransformEntry(FileSystemEntry& entry)
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext()
```

**Pull request merged**